### PR TITLE
Update gomod

### PR DESCRIPTION
**Pull request by MggMuggins**
I want to use the changes from https://github.com/canonical/lxd/pull/13213 in microcloud; however, building microcloud after `make update-gomod` fails due to the changes in https://github.com/canonical/lxd/commit/0d903b4037ed7b747f57efd1ad5171878eca7ea2. The changes in `internal/db` allow microcluster to compile again.

Alternatively, pass the context down through `DB.retry` as in https://github.com/MggMuggins/microcluster/commit/fde3e8629439f45175e9adbc4c6fd93d1c20437e. Let me know which is preferable.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -16,48 +18,79 @@ require (
...
 	github.com/kr/fs v0.1.0 // indirect
 	github.com/mattn/go-runewidth v0.0.15 // indirect
-	github.com/mattn/go-sqlite3 v1.14.19 // indirect
+	github.com/mattn/go-sqlite3 v1.14.22 // indirect
 	github.com/muhlemmer/gu v0.3.1 // indirect
+	github.com/natefinch/wrap v0.2.0 // indirect
+	github.com/oklog/ulid/v2 v2.1.0 // indirect
+	github.com/openfga/api/proto v0.0.0-20240318145204-66b9e5cb403c // indirect
+	github.com/openfga/language/pkg/go v0.0.0-20240328221832-7a66b52ab5fb // indirect
+	github.com/openfga/openfga v1.5.1 // indirect
```
Comment (markylaing) [inline go.mod:55]: A bit concerning that dependants of LXD now need to pull in OpenFGA and all of their associated libraries now. We'll need to sort this on the LXD side.

File: internal/db/db.go [hunk 2]
```diff
@@ -126,12 +126,12 @@ func (db *DB) Transaction(ctx context.Context, f func(context.Context, *sql.Tx)
...
 
-func (db *DB) retry(f func() error) error {
+func (db *DB) retry(f func(context.Context) error) error {
 	if db.ctx.Err() != nil {
-		return f()
+		return f(context.TODO())
 	}
 
-	return query.Retry(f)
+	return query.Retry(context.TODO(), f)
```
Comment (markylaing) [inline internal/db/db.go:134]: Now that these methods accept a context, we should pass the given context into the function wherever possible.

**Review by markylaing — CHANGES_REQUESTED** [review]
Can you please update this PR to pass down any context arguments wherever possible? Thanks.

File: go.mod (hunk 1 above)
Comment (tomponline) [reply go.mod:55]: yeah this needs sorting out before we can merge this.

Can you look into what is going on here and try to separate our openfga deps out a bit please?

**Comment by markylaing** [comment]
The reason for all of the OpenFGA imports can be shown with `go mod why -m github.com/openfga/openfga`. This gives:
```
# github.com/openfga/openfga
github.com/canonical/microcluster/example/api
github.com/canonical/lxd/lxd/response
github.com/canonical/lxd/lxd/db
github.com/openfga/openfga/pkg/storage
```

- The example microcluster uses `response.Response` to re-use LXD's response type.
- `lxd/response` imports `lxd/db` so that it can check for a sentinel error `db.ErrAlreadyDefined` when resolving a `SmartError`.
- `lxd/db` imports `openfga/pkg/storage` for the `OpenFGADatastore` implementation.

I'll try decoupling `lxd/response` from `lxd/db` as a first step. This package is used by microcluster, microcloud, and LXD cloud so it should be moved to `shared`.

**Comment by tomponline** [comment]
TIL `go mod why` thanks!

**Comment by markylaing** [comment]
> > I'll try decoupling lxd/response from lxd/db as a first step. This package is used by microcluster, microcloud, and LXD cloud so it should be moved to shared
> 
> Sounds good, perhaps we can use api.StatusErrorf(http.StatusConflict) instead of `db.ErrAlreadyDefined`.

Yeah that's my plan. Also, moving `lxd/response` to shared will be quite a headache I think. For now to move this PR along I'll just change the error but I'll create an issue to move it properly.

**Comment by tomponline** [comment]
@markylaing sounds good.

**Comment by markylaing** [comment]
With https://github.com/canonical/lxd/pull/13252 the go mod looks like:

```
module github.com/canonical/microcluster

go 1.22.0

replace github.com/canonical/lxd => github.com/markylaing/lxd v0.0.0-20240403093903-92aceaa46222

require (
	github.com/canonical/go-dqlite v1.21.0
	github.com/canonical/lxd v0.0.0-20240402145259-e51b0a89c188
	github.com/fsnotify/fsnotify v1.7.0
	github.com/google/renameio v1.0.1
	github.com/gorilla/mux v1.8.1
	github.com/olekukonko/tablewriter v0.0.5
	github.com/spf13/cobra v1.8.0
	golang.org/x/sys v0.18.0
	gopkg.in/yaml.v2 v2.4.0
)

require (
	github.com/Rican7/retry v0.3.1 // indirect
	github.com/armon/go-proxyproto v0.1.0 // indirect
	github.com/flosch/pongo2 v0.0.0-20200913210552-0d938eb266f3 // indirect
	github.com/fvbommel/sortorder v1.1.0 // indirect
	github.com/golang/protobuf v1.5.4 // indirect
	github.com/google/go-cmp v0.6.0 // indirect
	github.com/google/uuid v1.6.0 // indirect
	github.com/gorilla/schema v1.3.0 // indirect
	github.com/gorilla/securecookie v1.1.2 // indirect
	github.com/gorilla/websocket v1.5.1 // indirect
	github.com/gosexy/gettext v0.0.0-20160830220431-74466a0a0c4a // indirect
	github.com/inconshreveable/mousetrap v1.1.0 // indirect
	github.com/kr/fs v0.1.0 // indirect
	github.com/kr/pretty v0.3.1 // indirect
	github.com/mattn/go-runewidth v0.0.15 // indirect
	github.com/mattn/go-sqlite3 v1.14.22 // indirect
	github.com/muhlemmer/gu v0.3.1 // indirect
	github.com/pkg/errors v0.9.1 // indirect
	github.com/pkg/sftp v1.13.6 // indirect
	github.com/pkg/xattr v0.4.9 // indirect
	github.com/rivo/uniseg v0.4.7 // indirect
	github.com/rogpeppe/go-internal v1.10.0 // indirect
	github.com/sirupsen/logrus v1.9.3 // indirect
	github.com/spf13/pflag v1.0.5 // indirect
	github.com/zitadel/oidc/v2 v2.12.0 // indirect
	golang.org/x/crypto v0.21.0 // indirect
	golang.org/x/net v0.22.0 // indirect
	golang.org/x/oauth2 v0.18.0 // indirect
	golang.org/x/sync v0.6.0 // indirect
	golang.org/x/term v0.18.0 // indirect
	golang.org/x/text v0.14.0 // indirect
	google.golang.org/appengine v1.6.8 // indirect
	google.golang.org/protobuf v1.33.0 // indirect
	gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c // indirect
	gopkg.in/square/go-jose.v2 v2.6.0 // indirect
)

```

**Comment by MggMuggins** [comment]
Thanks all; rebased and ready to go

**Review by masnax — APPROVED** [review]
Looks good thanks!

**Pull request merged**